operators/constellation-node-operator/internal/infrastructure/scalinggroup_controller.go

```diff
@@ -0,0 +1,137 @@
+	cspapi "github.com/edgelesssys/constellation/v2/operators/constellation-node-operator/v2/internal/cloud/api"
+	"github.com/edgelesssys/constellation/v2/operators/constellation-node-operator/v2/internal/executor"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ExternalScalingGroupReconciler reconciles on scaling groups in CSP infrastructure.
+// It does NOT reconcile on k8s resources.
+// Instead, it scans the cloud provider infrastructure and changes k8s resources accordingly.
+type ExternalScalingGroupReconciler struct {
+	// uid is the unique identifier of the Constellation cluster.
+	uid                    string
+	scalingGroupDiscoverer scalingGroupDiscoverer
+	k8sClient              k8sReadWriter
+}
+
+// NewExternalScalingGroupReconciler creates a new InfrastructureReconciler.
+func NewExternalScalingGroupReconciler(uid string, discoverer scalingGroupDiscoverer, k8sClient k8sReadWriter) *ExternalScalingGroupReconciler {
+	return &ExternalScalingGroupReconciler{
+		uid:                    uid,
+		scalingGroupDiscoverer: discoverer,
+		k8sClient:              k8sClient,
+	}
+}
+
+// Reconcile reconciles on scaling groups in CSP infrastructure.
+func (r *ExternalScalingGroupReconciler) Reconcile(ctx context.Context) (executor.Result, error) {
+	logr := log.FromContext(ctx)
+	logr.Info("reconciling external scaling groups")
+
+	nodeGroups, err := r.scalingGroupDiscoverer.ListScalingGroups(ctx, r.uid)
+	if err != nil {
+		return executor.Result{}, err
+	}
+
+	existingNodeGroups := map[string]struct{}{}
+
+	// create all scaling groups that are newly discovered
+	for _, group := range nodeGroups {
+		err := createScalingGroupIfNotExists(ctx, newScalingGroupConfig{
+			k8sClient:            r.k8sClient,
+			resourceName:         group.Name,
+			groupID:              group.GroupID,
+			nodeGroupName:        group.NodeGroupName,
+			autoscalingGroupName: group.AutoscalingGroupName,
+			role:                 group.Role,
+		})
+		if err != nil {
+			return executor.Result{}, err
+		}
+		existingNodeGroups[group.Name] = struct{}{}
+	}
+
+	logr.Info("ensured scaling groups are created", "count", len(nodeGroups))
+
+	// delete all scaling groups that no longer exist
+	var scalingGroups updatev1alpha1.ScalingGroupList
+	if err := r.k8sClient.List(ctx, &scalingGroups); err != nil {
+		return executor.Result{}, err
+	}
+	for _, group := range scalingGroups.Items {
+		if _, ok := existingNodeGroups[group.Name]; !ok {
+			logr.Info("deleting scaling group", "name", group.Name)
+			err := r.k8sClient.Delete(ctx, &group)
+			if err != nil {
+				return executor.Result{}, err
+			}
+		}
+	}
+
+	logr.Info("external scaling groups reconciled")
+
+	return executor.Result{}, nil
+}
+
+func createScalingGroupIfNotExists(ctx context.Context, config newScalingGroupConfig) error {
```

Comment:
just a nit how we could ease unit testing in the future IMHO. I'd separate the k8s client from the config and restrain the interface to only Create(). 
Could have saved you some trouble writing the complete fake. But would just keep as is now

Reply:
Agreed. If we touch this in the future, this would be a good improvement.